File: 2019/kotlin/src/main/kotlin/day01.kt
import kotlin.math.floor

fun calculateFuel(mass: Int) = floor(mass.toDouble() / 3.0).toInt() - 2

fun calculateFuelRec(mass: Int): Int {
    val fuel = calculateFuel(mass)
    return if (fuel < 0) 0
    else fuel + calculateFuelRec(fuel)
}

fun main(args: Array<String>) {
    val input = readLines("day01.txt") { it.toInt() }

    println("Part one: ${input.sumBy { calculateFuel(it) }}")
    println("Part two: ${input.sumBy { calculateFuelRec(it) }}")
}

File: 2015/rust/src/day01.rs
use std::fs;

fn part_one(input: &str) -> i32 {
    input.chars().fold(0, |acc, c| match c {
        '(' => acc + 1,
        ')' => acc - 1,
        _ => acc,
    })
}

fn part_two(input: &str) -> usize {
    input
        .chars()
        .scan(0, |floor, c| {
            *floor += match c {
                '(' => 1,
                ')' => -1,
                _ => 0,
            };
            if *floor == -1 { None } else { Some(*floor) }
        })
        .count()
        + 1
}

fn main() {
    let input = fs::read_to_string("input/day01.txt").expect("Failed to read input file");

    println!("Part one: {}", part_one(&input));
    println!("Part two: {}", part_two(&input));
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_part_one() {
        assert_eq!(part_one("(())"), 0);
        assert_eq!(part_one("()()"), 0);
        assert_eq!(part_one("((("), 3);
        assert_eq!(part_one("(()(()("), 3);
        assert_eq!(part_one("))((((("), 3);
        assert_eq!(part_one("())"), -1);
        assert_eq!(part_one("))("), -1);
        assert_eq!(part_one(")))"), -3);
        assert_eq!(part_one(")())())"), -3);
    }

    #[test]
    fn test_part_two() {
        assert_eq!(part_two(")"), 1);
        assert_eq!(part_two("()())"), 5);
    }
}
